packages/hooks/src/Provider.tsx

```diff
@@ -0,0 +1,94 @@
+import WalletConnectProvider from '@walletconnect/web3-provider';
+import Web3Modal from 'web3modal';
+
+export interface Web3ContextType {
+  connectWallet: (() => void) | undefined;
+  signer: JsonRpcSigner | null | undefined;
+  userAddress: string | null | undefined;
+  disconnectWallet: (() => void) | undefined;
+  network: string;
+  chainId: number | undefined;
+  connected: boolean;
+  provider: ethers.providers.Web3Provider | null | undefined;
+}
+
+export const Web3Context = React.createContext<Web3ContextType | undefined>(undefined);
+
+export interface ProviderProps {
+  /**
+   * @dev The network you want to connect to.
+   * @example 'mainnet'
+   * @type string
+   */
+  network: string;
+  /**
+   * @dev Your Infura project ID. This is required if you want to support WalletConnect.
+   * @type string
+   */
+  infuraId?: string;
+}
+
+/**
+ * @dev The global provider that handles and stores all the web3 connections. Wrap your entire App with this component.
+ * @param children Your app.
+ * @param network The network you want to connect to.
+ * @param infuraId Your Infura project ID. This is required if you want to support WalletConnect.
+ */
+export const Provider: React.FC<ProviderProps> = ({ children, network, infuraId }) => {
+  const [signer, setSigner] = React.useState<null | JsonRpcSigner>(null);
+  const [provider, setProvider] = React.useState<null | ethers.providers.Web3Provider>(null);
+  const [userAddress, setUserAddress] = React.useState<null | string>(null);
+  const [web3Modal, setWeb3Modal] = React.useState<null | Web3Modal>();
+  const [chainId, setChainId] = React.useState<number>();
+  const [connected, setConnected] = React.useState<boolean>(false);
+
+  const connectWallet = React.useCallback(async () => {
+    console.log('connectWallet');
+    const web3Modal = new Web3Modal({
+      network,
+      providerOptions: {
+        walletconnect: {
+          package: WalletConnectProvider,
+          options: {
+            infuraId,
+          },
+        },
+      },
+    });
+    const connection = await web3Modal.connect();
+    const provider = new ethers.providers.Web3Provider(connection);
+    const signer = provider.getSigner();
+    setWeb3Modal(web3Modal);
+    setSigner(signer);
+    setProvider(provider);
+    setUserAddress(await signer.getAddress());
+    setChainId(await signer.getChainId());
+    setConnected(true);
+  }, [Web3Modal, web3Modal, WalletConnectProvider, network, infuraId, ethers]);
+
+  const disconnectWallet = React.useCallback(async () => {
+    web3Modal?.clearCachedProvider();
+    setSigner(null);
+    setUserAddress(null);
+    setConnected(false);
+  }, [web3Modal]);
+
+  const value = React.useMemo(
+    () => ({
+      connectWallet,
+      signer,
+      userAddress,
+      disconnectWallet,
+      connected,
+      provider,
+      network,
+      chainId,
+    }),
+    [connectWallet, signer, userAddress, web3Modal, connected, provider, network, chainId]
+  );
```

Comment:
`useMemo` probably isn't helping in the way you think it is.

Some resources that might explain why/provide alternatives:

- https://github.com/kentcdodds/old-kentcdodds.com/blob/319db97260078ea4c263e75166f05e2cea21ccd1/content/blog/how-to-optimize-your-context-value/index.md
- https://github.com/facebook/react/issues/15156#issuecomment-474590693
- https://kentcdodds.com/blog/how-to-optimize-your-context-value

I think we should just go with this for now and figure out a better pattern later, just wanted to throw that out there :)

Reply:
Thanks for sharing these resources. Super helpful